### ISSUE TITLE
Houdini: workfile tool doesn't switch context properly

### ISSUE THREAD

**Issue by MustafaJafar** [ynput/ayon-core#180]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior:

Sometimes context is not changed when changing context with workfile tool

![image](https://github.com/ynput/ayon-core/assets/20871534/1069acdb-d17c-417a-91ed-dd4b8c6df78e)

### Expected Behavior:

It should change the asset context.

### Version

Launcher `1.0.3-dev.1`
Ayon Core `0.3.0-dev.1`
Ayon Server `1.0.4+202402231406`

### What platform you are running on?

Windows

### Steps To Reproduce:

1. Open Houdini 
2. Use workfile tool to open a workfile of another asset

### Are there any labels you wish to add?

- [X] I have added the relevant labels to the bug report.

### Additional context:

I'm not sure if this problem happens because I'm constantly testing new PRs and updating my `ayon-core` clone.

[cuID:[AY-3851](https://app.clickup.com/t/86bxwjrxy)]

**Comment by BigRoy** [comment]
I wonder if this may have to do with it relying on the "context" saved in the scene - and not necessarily the context of that asset. E.g. when saving to different assets in the first place:

https://github.com/ynput/ayon-core/blob/d724a666331b042d9908a1c6e4a4271348174c76/client/ayon_core/hosts/houdini/api/pipeline.py#L171-L175

It seems to in the menu show the `os.environ["AYON_FOLDER_PATH"]` so when switched it should've updated that env var.
Might be good to just test and print that environment variable to see whether it has updated since - and if not, then we know at least the bug isn't related to the Houdini menu itself not refreshing. ;)

Do you get any logged/printed errors when switching / opening the other that might be related to updating of the context failing?

**Comment by MustafaJafar** [comment]
> You mention "sometimes" - but have you found a way to reliably reproduce it? Maybe it always occurs on your end if you switch from one particular asset to another particular asset.

You are correct.. It's happening when switching to a particular asset.
any ideas how to examine the workfile, maybe it's corrupted.

**Comment by BigRoy** [comment]
You could try running that `get_context_data` call on Houdini host in that scene file to see if it does anything 'wrong' in what it returns.

Also, print `os.environ["AYON_FOLDER_PATH"]` to confirm whether that's indeed invalid.

And as stated, of course keep an eye out on the console to see if there are any errors whatsoever being reported.

**Comment by MustafaJafar** [comment]
TL;DR 
So, This is confusing (the issue was solved when publishing the workfile). 

---
In one asset 
`registered_host().get_context_data()["workfile"]["folderPath"]` was correct but  `os.environ["AYON_FOLDER_PATH"]` didn't update properly on context change.

In another asset  `os.environ["AYON_FOLDER_PATH"]` was correct but `registered_host().get_context_data()["workfile"]["folderPath"]` didn't match because of the letter case! 

However, both problems were solved when publishing the workfile. 
It's fair to close this issue now but I'm still confused 😅.

**Comment by BigRoy** [comment]
Ok, so I can reproduce this problem and I must admit - it's quite a dangerous one.

1. Take one workfile. (e.g. `char/model`)
2. Save it into another task.  (e.g. `char/look`)
3. Then open the the original workfile (e.g. from a new session or same session).  (e.g. `char/model`)
4. Then open the workfile that was saved to the other task -  (e.g. `char/look`)
5. Now after opening the `char/look` context the current context will still be `char/model` because it reads that from the workfile.


So after, with workfiles tool, opening the file in `char/look` the context still is `char/model`
```python
os.environ["AYON_TASK_NAME"]
```
And the environment variable then also still points to the model.

In this case I'm showing the issue with task name, but I believe it's also reproducible with folder path since it likely reads the "current context" from the context stored in the workfile.

@iLLiCiTiT @antirotor this seems like unexpected behavior - can you confirm?

**Comment by MustafaJafar** [comment]
I just double checked the issue and the bug has gone. 

The bug originally happened when saving workfile to another context and closing the workfile without saving. 
Steps: 
1. Create two empty new folders/assets
2. open the first folder from launcher, save, open second folder from workfile tool, save from the workfile tool and close houdini without saving manualy or hitting `ctrl+s`.
3. from launcher open the second folder (it should be cool)
4. from workifle tool: open first folder (should be cool), open second folder again (the error ocurrs) 

This scenario never happened with the merged PR.
